### ISSUE TITLE
Upgrading vmstorage up to 1.97 or later (including the latest release 1.103.0) causes search performance degradation

### ISSUE THREAD

**Issue by mvanyushkin** [VictoriaMetrics/VictoriaMetrics#7009]
### Describe the bug

Hello everyone!
We had a VM cluster v1.93.8. Then, for some reasons our production cluster was supposed to be updated up to the closest LTS version 1.97.1. 
We did some preparing procedure and finally got the updated cluster up to 1.97.1 (vminsert, vminsert, vmstorage as well).
Almost immediately we noticed that pql search queries handled by the cluster could take some more time than we used to.

After some research we've figured out that the  "Request Durations p50" metric has dramatically changed
<img width="616" alt="p50select" src="https://github.com/user-attachments/assets/73e9a6a1-3e07-4369-ad9b-9bcbfbfbd1c5">

Going deeper an anomaly has been found - cache hit ratio of "indexdb/tagFiltersToMetricIDs" has changed after upgrading to 1.97.1
<img width="408" alt="cache_hit" src="https://github.com/user-attachments/assets/46b2893f-1d86-4585-aa44-c01f59aa7c69">
<img width="573" alt="cpu-usage" src="https://github.com/user-attachments/assets/46a1a66e-1a1e-4562-a862-9d6eaa4df1b1">

so seemingly cache under usage could be a root cause of search performance degradation.

moreover, taking pprof profile we get the following picture (it could indirectly show cache misses)
v1.96.0
~~~
      flat  flat%   sum%        cum   cum%
     9.62s 14.88% 14.88%      9.62s 14.88%  cmpbody
     4.93s  7.62% 22.50%      4.97s  7.69%  runtime.cgocall
     4.19s  6.48% 28.98%      4.27s  6.60%  runtime.mapaccess1_fast64
     3.16s  4.89% 33.86%      8.29s 12.82%  github.com/VictoriaMetrics/VictoriaMetrics/lib/blockcache.(*cache).GetBlock
     3.14s  4.86% 38.72%      3.14s  4.86%  runtime/internal/syscall.Syscall6
     2.77s  4.28% 43.00%      6.15s  9.51%  github.com/VictoriaMetrics/fastcache.(*bucket).Get
     2.23s  3.45% 46.45%      6.65s 10.28%  github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.binarySearchKey
     1.75s  2.71% 49.16%      1.75s  2.71%  runtime.memmove
     1.42s  2.20% 51.35%      3.85s  5.95%  github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*partSearch).Seek.func2
     1.36s  2.10% 53.46%     26.51s 40.99%  github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*partSearch).Seek
~~~

v1.97.0
~~~
      flat  flat%   sum%        cum   cum%
    17.20s 15.88% 15.88%     17.20s 15.88%  cmpbody
     8.97s  8.28% 24.17%      9.07s  8.38%  runtime.cgocall
     6.64s  6.13% 30.30%     17.12s 15.81%  github.com/VictoriaMetrics/VictoriaMetrics/lib/blockcache.(*cache).GetBlock
     6.12s  5.65% 35.95%      6.33s  5.85%  runtime.mapaccess1_fast64
     4.07s  3.76% 39.71%     11.84s 10.93%  github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.binarySearchKey
     3.46s  3.20% 42.90%      3.46s  3.20%  runtime/internal/syscall.Syscall6
     3.10s  2.86% 45.77%      7.47s  6.90%  github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*partSearch).Seek.func2
     2.88s  2.66% 48.43%      6.78s  6.26%  github.com/VictoriaMetrics/fastcache.(*bucket).Get
     2.67s  2.47% 50.89%      2.67s  2.47%  runtime.memmove
     2.66s  2.46% 53.35%      3.97s  3.67%  runtime.mapaccess1
~~~

Thank you for your job :)    

### To Reproduce

1. get VM cluster with the workload: 
- 1k search rps totally 
- 160k of ingestion point per vmstorage 
- 5m of active timeseries per vmstorage

2. use version 1.97.0 or later

### Version

vmstorage-20240201-133818-tags-v1.97.1-cluster-0-g765917ae6

### Additional information

we've checked a bunch of versions:

- 1.93-1.96 are NOT affected
- 1.97, 1.102, 1.103 are affected

**Comment by rtm0** [comment]
@mvanyushkin Please help to understand your workload more precisely:

> 1k search rps totally

Does this total include only select API calls (such as `/api/v1/query`) or also include insert (such as `/api/v1/write`)? 

> 160k of ingestion point per vmstorage

This is the `Datapoints ingestion rate`, correct?

**Comment by mvanyushkin** [comment]
### Search load

Request rate
<img width="861" alt="image" src="https://github.com/user-attachments/assets/8725bb9f-9468-4bb7-bb16-63a5f9524f37">

<img width="859" alt="image" src="https://github.com/user-attachments/assets/d3a64d88-0ae1-4af4-9816-486bffda1384">



### Ingestion Load

<img width="431" alt="image" src="https://github.com/user-attachments/assets/69aa4a2f-43ca-4463-89a2-abae5d5bb779">

<img width="856" alt="image" src="https://github.com/user-attachments/assets/422c9d86-16bd-4c9b-8bca-7b8d0a6555fa">

### Caches usage
indexdb/tagFiltersToMetricIDs
<img width="850" alt="image" src="https://github.com/user-attachments/assets/a5ed72f2-c749-4f00-ba27-ea0775076d3c">
storage/indexBlocks
<img width="856" alt="image" src="https://github.com/user-attachments/assets/9cce5974-d5df-45bc-84f7-4c88506e73ff">

**Comment by rtm0** [comment]
Hi @mvanyushkin, thank you for sharing those charts. I've been trying to reproduce it locally, but for now not getting the desired outcome. That does not mean that there is no issue, but I'd like to ask for more data. Would it be a problem for you to share the entire `VictoriaMetrics - cluster` dashboard for 9/12 - 9/15 period? Grafana allows that (there is a `Share` button that uploads the dashboard to some public resource. They say it tries to remove some sensitive data but I am not sure to which extent. Feel free to say no if you think that some sensitive may be exposed unintentionally.

**Comment by mvanyushkin** [comment]
Thank you for supporting the issue we've faced with!
Sure! The snapshot is available by the link https://snapshots.raintank.io/dashboard/snapshot/7lc8Jo7KMN9Rn90clpxwyxh9YY6ZlKEP?orgId=0
I have to say the dashboard our team uses is a bit customized :) But either way it's based on the standard Victoria Metrics Cluster dashboard. 
I striped out sensitive data as much as I could and changed the interval value from 1m to 10m (in other way raintank didn't accept the snapshot due to too large size).
Hope the snapshot is informative for you.
And I'm going to share with you some more workflow details:
- **9/12 at 11:30 AM:** a part of ingestion load was moved from the current cluster (under maintenance) to another one (in order to have the upgrade process easier having less high cardinality data). 
  - Request Duration p50 decreased from 100-200ms to 30-50ms. 
  - Probably it happened due to search traffic was moved into another cluster as well.
- **9/12 at 01:00PM:** did vmstorage rolling update up to the 1.97.1. 
  - Request Duration p50 increased x5, from 100-200ms to 1...2 seconds.
- **9/13 at 01:30PM** the part of ingestion load was moved back to the current cluster.
  - Request Duration p50 increased from 30-50ms to 2..5 seconds.
- **9/14 at 6:30AM** Request Duration p50 significantly increased to 4-6 second. I guess it happened because of some our clients started retrying slow search requests.
- **9/14 from 09:00AM to 11:30AM:** rolling back vmstorage from 1.97.1 to 1.93.8 
  - Request Duration p50 decreased from 4-6 seconds to 30..50ms (which is a typical value for weekends)

**NB: vmselect/vminsert stay on 1.97.1 all the time**

<img width="857" alt="image" src="https://github.com/user-attachments/assets/c8a80750-910b-4a88-aadc-da2d2658a2d5">

I could provide any other details if you need :) 
Thank you!

**Comment by rtm0** [comment]
Thanks @mvanyushkin! This should be very helpful. I am still exploring the dashboard that you've shared and will definitely look at profiles. In the mean time just wanted to know if you tried:

1.  Upgrading without prior traffic removal?
2.  Performing a canary deployment (when only a small part of vmstorage replicas is updated) and then comparing the behavior of two versions running simultaneously?

If so, what had been the results?

**Comment by mvanyushkin** [comment]
Yes, we did :)
Actually, we tried updating 2 clusters of VictoriaMetrics in different datacenters and got absolutely the same behaviour.
Hm..... I'm going to put you in the picture a bit deeper than I was going :) 
Alright, we have 2 datacenters, let's call them A and B.

- VM in A ingests ~6mil datapoints per second
- VM in B ingests ~2.7 datapoints per second

![image](https://github.com/user-attachments/assets/9738dd63-3b36-4fb1-810a-c8e522d31eef)
**The snapshot I've shared with you above describes the datacenter A.**

### The full timeline is:

- 8/29 a vmstorage node was updated in the DC B to 1.97.1 (**only one node, NO TRAFFIC REMOVAL**)
...
- 9/11 rolling update DC B vmstorages to 1.97.1 (ROLLING MODE, NO TRAFFIC REMOVAL)
.....
- 9/12 rolling update DC A storages to 1.97.1 (rolling mode, having done traffic removal prematurely)
- 9/13 noticed significant performance degradation datacenter A (after returning traffic). And also we did pay attention to datacenter B and found the same problem (in less
- 9/14 we made a decision to roll vmstorages back to 1.93.8 in both datacenters and search perf degradation has gone.
Datacenter B - vmselect Request Duration p50
![image](https://github.com/user-attachments/assets/a044ee76-c44a-4a10-b2d0-3b3f4d3e9f03)
Datacenter B - vmstorages CPU consumption
![image](https://github.com/user-attachments/assets/10df767c-dc27-40cf-9511-2b57d82640ff)
Datacenter B
![image](https://github.com/user-attachments/assets/3961e479-024e-4fbe-bc4e-33a0481e4d36)
Datacenter B - indexdb/tagFiltersToMetricIDs cache hit ratio
![image](https://github.com/user-attachments/assets/0086aeca-9da8-4c63-be26-ea34151ea6d4) 
Datacenter B - storage/indexBlocks cache hit ratio for node 12
![image](https://github.com/user-attachments/assets/79e1977a-25c7-4664-95b3-e57278e6ff59)



The main mistake from our side was to be focused on the data ingestion during update :( 
We didn't pay enough attention to the vmselect metrics and noticed degradation too late (when the second cluster was updated).

**Comment by rtm0** [comment]
Hi @mvanyushkin, 

One hypothesis that we've got it that the culprit could be the tagFiltersToMetricIDs cache value compression that has been introduced in v1.97.x. We are unable to prove it locally though.

Would it be possible for you run it for us? I believe updating one vmstorage replica should be enough to see the difference (similarly to what you've done in DC B, see `Datacenter B - indexdb/tagFiltersToMetricIDs cache hit ratio` chart from your previous post)

The image is here: docker.io/victoriametrics/vmstorage:remotes-rtm0-v1.97.1-cluster-issue-7009-0-g7f636c7e89

This version only adds this [patch](https://github.com/rtm0/VictoriaMetrics/commit/7f636c7e893a19a0be5153dd779192ea3f7fc95b) on top of the original v1.97.1.  

Thanks!

cc: @valyala

**Comment by mvanyushkin** [comment]
I've already done some experiments for you!
I built vmstorage from sources and took a vmstorage node for experiments from datacenter B.
I did some git bisect session and I can prove that the problem comes with the following change https://github.com/VictoriaMetrics/VictoriaMetrics/commit/6c214397eda2fd77c348aff28e391e4703c215db

![image](https://github.com/user-attachments/assets/69245b42-4317-4e31-a13a-b831cc2eeeca)
![image](https://github.com/user-attachments/assets/890dc6a0-1211-4eb3-acf6-adfb720508d3)

**Comment by mvanyushkin** [comment]
[rtm0](https://github.com/rtm0) hi!
I've just tried your patch
![image](https://github.com/user-attachments/assets/f62e6123-0428-4370-81d5-7c45885a9147)
seems the situation is the same
![image](https://github.com/user-attachments/assets/476f6e58-2221-46c8-888a-e7d6f651fed0)

**Comment by rtm0** [comment]
Hey @mvanyushkin, fantastic work with bisecting! What is interesting is that the patch that I've sent you attempts to revert the very change that you've identified :D Obviously it did not work based on your last reply. But I will take another look at that commit to see if I've missed something. I will send you another image to try.

Once again, thank you for actively participating in investigation. That really helps, provided that we are unable to reproduce the issue locally!

**Comment by valyala** [comment]
FYI, the attached CPU profiles at https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7009#issuecomment-2358221948 helped detecting and fixing an additional performance bottleneck at `indexSearch.containsTimeRange()` function - see 787b9cd9a09f49ace10f23e1273eb7e21a3c463d for details. This commit will be included in the next release.

Also @rtm0 should provide a bugfix for the issue introduced in the commit  https://github.com/VictoriaMetrics/VictoriaMetrics/commit/6c214397eda2fd77c348aff28e391e4703c215db soon. This commit broke caching for [label filters](https://docs.victoriametrics.com/keyconcepts/#filtering), which match zero [time series](https://docs.victoriametrics.com/keyconcepts/#time-series) on the selected time range. The bugfix will be also included in the next release.

**Comment by mvanyushkin** [comment]
That's awesome news!
I'm looking forward to get the bugfix! :)
Thank you guys! ❤️

**Comment by rtm0** [comment]
@mvanyushkin we have merged the fix into the cluster branch. Could you please build the vmstorage image from it and try? Thank you!

**Comment by mvanyushkin** [comment]
> @mvanyushkin we have merged the fix into the cluster branch. Could you please build the vmstorage image from it and try? Thank you!

Great job, guys!!! 
I confirm the issue has gone!
p50
![image](https://github.com/user-attachments/assets/3c94e788-8d9f-4032-9247-95c19b11f11e)
p99
![image](https://github.com/user-attachments/assets/b1c62e44-b299-43c6-8f75-0e501e40ef97)
cpu
![image](https://github.com/user-attachments/assets/dbf04115-106f-4c8d-a7d6-7c080ef58905)

And the actual pprof profile if you are interested in [origin-cluster-vmstorage.pprof.gz](https://github.com/user-attachments/files/17078958/origin-cluster-vmstorage.pprof.gz)

Thank you very much guys! The issue can be closed!
I would buy you a beer if I could! 🍺

**Comment by hagen1778** [comment]
The bugfix was included into [v1.104.0](https://docs.victoriametrics.com/changelog/#v11040), [v1.102.3](https://docs.victoriametrics.com/changelog/#v11023) and [v1.97.8](https://docs.victoriametrics.com/changelog/#v1978).
